search/src/kernels.cu

```diff
@@ -184,180 +186,427 @@ extern "C" void devicePool(int sourceWidth, int sourceHeight, float *source,
-		deviceTests, deviceSearchResults, deviceImgTimes);
-
-	// Read back results
-	checkCudaErrors(cudaMemcpy(bestTrajects, deviceSearchResults,
-				sizeof(trajectory)*resultsCount, cudaMemcpyDeviceToHost));
-
-	checkCudaErrors(cudaFree(deviceTests));
-	checkCudaErrors(cudaFree(deviceImgTimes));
-	checkCudaErrors(cudaFree(deviceSearchResults));
-	checkCudaErrors(cudaFree(devicePsiPhi));
+    // Allocate Device memory
+    trajectory *deviceTests;
+    float *deviceImgTimes;
+    float *devicePsiPhi;
+    trajectory *deviceSearchResults;
+
+    checkCudaErrors(cudaMalloc((void **)&deviceTests, sizeof(trajectory)*trajCount));
+    checkCudaErrors(cudaMalloc((void **)&deviceImgTimes, sizeof(float)*imageCount));
+    checkCudaErrors(cudaMalloc((void **)&devicePsiPhi,
+        sizeof(float)*psiPhiSize));
+    checkCudaErrors(cudaMalloc((void **)&deviceSearchResults,
+        sizeof(trajectory)*resultsCount));
+
+    // Copy trajectories to search
+    checkCudaErrors(cudaMemcpy(deviceTests, trajectoriesToSearch,
+            sizeof(trajectory)*trajCount, cudaMemcpyHostToDevice));
+
+    // Copy image times
+    checkCudaErrors(cudaMemcpy(deviceImgTimes, imageTimes,
+            sizeof(float)*imageCount, cudaMemcpyHostToDevice));
+
+    // Copy interleaved buffer of psi and phi images
+    checkCudaErrors(cudaMemcpy(devicePsiPhi, interleavedPsiPhi,
+        sizeof(float)*psiPhiSize, cudaMemcpyHostToDevice));
+
+    //dim3 blocks(width,height);
+    dim3 blocks(width/THREAD_DIM_X+1,height/THREAD_DIM_Y+1);
+    dim3 threads(THREAD_DIM_X,THREAD_DIM_Y);
+
+
+    // Launch Search
+    searchImages<<<blocks, threads>>> (trajCount, width,
+        height, imageCount, minObservations, devicePsiPhi,
+        deviceTests, deviceSearchResults, deviceImgTimes);
+
+    // Read back results
+    checkCudaErrors(cudaMemcpy(bestTrajects, deviceSearchResults,
+                sizeof(trajectory)*resultsCount, cudaMemcpyDeviceToHost));
+
+    checkCudaErrors(cudaFree(deviceTests));
+    checkCudaErrors(cudaFree(deviceImgTimes));
+    checkCudaErrors(cudaFree(deviceSearchResults));
+    checkCudaErrors(cudaFree(devicePsiPhi));
 }
 
+/*
+ * Searches through images (represented as a flat array of floats) looking for most likely
+ * trajectories in the given list. Outputs a results image of best trajectories. Returns a
+ * fixed number of results per pixel specified by RESULTS_PER_PIXEL
+ * filters results using a sigmaG-based filter and a central-moment filter.
+ */
+__global__ void searchFilterImages(int trajectoryCount, int width, int height,
+        int imageCount, int minObservations, float *psiPhiImages,
+        trajectory *trajectories, trajectory *results, float *imgTimes,
+        float sGL0, float sGL1, float sigmaGCoeff, float minLH)
+{
+
+    // Get trajectory origin
+    const unsigned short x = blockIdx.x*THREAD_DIM_X+threadIdx.x;
+    const unsigned short y = blockIdx.y*THREAD_DIM_Y+threadIdx.y;
+    float lcArray[MAX_NUM_IMAGES];
```

Comment:
These arrays consume a lot of thread local memory (looks like 100 * 4 * 4 bytes), which limits the number of threads that can be simultaneously active on each of the GPUs multiprocessors. Having more threads active on each multiprocessor allows more memory bandwidth to be used.
I'm not sure if you've noticed a significant performance impact when using these large local arrays, but it seems to me that you would have to reduce the number of threads per block, and would see quite a bit less performance.
If you've verified the results are correct and the performance is good, then no need to think too much about this! Just something to be aware of as a general GPU performance tip.
At the very least it might be worth it to try setting the MAX_NUM_IMAGES to the minimum you need and see if there is a noticeable performance difference.